staking/integration-tests/tests/stability.rs

```diff
@@ -0,0 +1,833 @@
+        Arbitrary,
+        Gen,
+        QuickCheck,
+    },
+    solana_sdk::{
+        pubkey::Pubkey,
+        signature::Keypair,
+        signer::Signer,
+    },
+    staking::{
+        error::ErrorCode as StakingErrorCode,
+        state::positions::{
+            Position,
+            PositionState,
+            Target,
+            TargetWithParameters,
+        },
+    },
+    std::{
+        cmp::min,
+        collections::HashMap,
+        convert::TryInto,
+        fmt::Debug,
+    },
+};
+
+// These constants define the parameters of the stability test.
+const MAX_DELEGATION_AMOUNT: u64 = 2 * FRAC_64_MULTIPLIER;
+const MAX_UNDELEGATION_AMOUNT: u64 = FRAC_64_MULTIPLIER / 10;
+const NUM_DELEGATORS: u64 = 1_000;
+const NUM_OPERATIONS: u64 = 10_000;
+const MAX_CAP: u64 = 100 * FRAC_64_MULTIPLIER;
+const Y: u64 = FRAC_64_MULTIPLIER / 10 / 52;
+
+#[derive(Debug)]
+struct KeypairWrapper(Keypair);
+
+impl Clone for KeypairWrapper {
+    fn clone(&self) -> Self {
+        KeypairWrapper(self.0.insecure_clone())
+    }
+}
+
+#[derive(Debug, Clone)]
+enum Operation {
+    Delegate {
+        publisher: usize,
+        delegator: usize,
+        amount:    u64,
+    },
+    Undelegate {
+        publisher: usize,
+        delegator: usize,
+        amount:    u64,
+    },
+    AdvanceDelegationRecord {
+        publisher: usize,
+        delegator: usize,
+    },
+    Advance {
+        cap: u64,
+    },
+    Slash {
+        publisher: usize,
+        amount:    u64,
+    },
+    MergePositions {
+        publisher: usize,
+        delegator: usize,
+    },
+    CreateGovernancePosition {
+        delegator: usize,
+        amount:    u64,
+    },
+    CloseGovernancePosition {
+        delegator: usize,
+        amount:    u64,
+    },
+}
+
+impl Operation {
+    fn get_name(&self) -> String {
+        match self {
+            Operation::Delegate { .. } => "Delegate".to_string(),
+            Operation::Undelegate { .. } => "Undelegate".to_string(),
+            Operation::AdvanceDelegationRecord { .. } => "AdvanceDelegationRecord".to_string(),
+            Operation::Advance { .. } => "Advance".to_string(),
+            Operation::Slash { .. } => "Slash".to_string(),
+            Operation::MergePositions { .. } => "MergePositions".to_string(),
+            Operation::CreateGovernancePosition { .. } => "CreateGovernancePosition".to_string(),
+            Operation::CloseGovernancePosition { .. } => "CloseGovernancePosition".to_string(),
+        }
+    }
+}
+
+#[derive(Clone)]
+struct StabilityTestProps {
+    operations: Vec<Operation>,
+    publishers: Vec<KeypairWrapper>,
+    delegators: Vec<KeypairWrapper>,
+}
+
+impl Debug for StabilityTestProps {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("StabilityTestProps")
+    }
+}
+
+impl Arbitrary for StabilityTestProps {
+    fn arbitrary(g: &mut Gen) -> Self {
+        let publishers: Vec<KeypairWrapper> = (0..(MAX_PUBLISHERS - 10))
```

Comment:
why minus 10 here

Reply:
just because it's more realistic not to have all the publishers set.